### ISSUE TITLE
unable too handle off canvas pixels from aseprite

### ISSUE THREAD

**Issue by Hellzbellz123** [bikeshedder/sprity#4]
recently decided too switch too bevy-aseprite-ultra so i can manage fewer assets and a few of the .aseprite files seemed too cause crashes.

after much confusion i redid a test file and noticed that aseprite saves the pixels that get moved off canvas unless you tick a checkbox in the canvas pane. i moved the pixels back on canvas and the crashes dissappeared. 

is there an easy way too tell if this is the case inside an aseprite file?

```
thread 'IO Task Pool (2)' panicked at /home/hellzbellz/.local/share/cargo/registry/src/index.crates.io-6f17d22bba15001f/aseprite-loader-0.3.1/src/loader/mod.rs:246:36:
range end index 267512 out of range for slice of length 16384
stack backtrace:
   0:     0x77907cc88a7a - std::backtrace_rs::backtrace::libunwind::trace::h412452c983d651b0
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/../../backtrace/src/backtrace/libunwind.rs:116:5
   1:     0x77907cc88a7a - std::backtrace_rs::backtrace::trace_unsynchronized::ha8a3c781b65667e3
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x77907cc88a7a - std::sys::backtrace::_print_fmt::ha99311aa022c4ece
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/sys/backtrace.rs:66:9
   3:     0x77907cc88a7a - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hb0397a16f8a4e2cf
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/sys/backtrace.rs:39:26
   4:     0x77907ccd9043 - core::fmt::rt::Argument::fmt::he85d508f0ddcad01
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/fmt/rt.rs:177:76
   5:     0x77907ccd9043 - core::fmt::write::hf02ed2b042d806f2
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/fmt/mod.rs:1186:21
   6:     0x77907cc7c573 - std::io::Write::write_fmt::ha711cb2eb6205a4f
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/io/mod.rs:1839:15
   7:     0x77907cc888c2 - std::sys::backtrace::BacktraceLock::print::hc94b705f39fad33b
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/sys/backtrace.rs:42:9
   8:     0x77907cc8aeda - std::panicking::default_hook::{{closure}}::he365bc97e1cb4f6d
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/panicking.rs:268:22
   9:     0x77907cc8ad20 - std::panicking::default_hook::h78ee557e1864928e
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/panicking.rs:295:9
  10:     0x77907cc8b577 - std::panicking::rust_panic_with_hook::h3cca1f11c9c514e6
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/panicking.rs:801:13
  11:     0x77907cc8b40a - std::panicking::begin_panic_handler::{{closure}}::h89684e55fcb5b494
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/panicking.rs:674:13
  12:     0x77907cc88f59 - std::sys::backtrace::__rust_end_short_backtrace::h8c330503360e1342
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/sys/backtrace.rs:170:18
  13:     0x77907cc8b09c - rust_begin_unwind
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/std/src/panicking.rs:665:5
  14:     0x77907ccd5450 - core::panicking::panic_fmt::h324ae0e7c6fba0a6
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/panicking.rs:74:14
  15:     0x77907ccdbad7 - core::slice::index::slice_end_index_len_fail_rt::haa63ed26452da749
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/slice/index.rs:64:5
  16:     0x77907ccdbad7 - core::slice::index::slice_end_index_len_fail::h8083b1ad6e5b4930
                               at /rustc/e7c0d2750726c1f08b1de6956248ec78c4a97af6/library/core/src/slice/index.rs:58:5
  17:     0x575ca137eb70 - aseprite_loader::loader::AsepriteFile::combined_frame_image::h4eeb46e484c86267
  18:     0x575ca133e77d - <bevy_aseprite_ultra::loader::AsepriteLoader as bevy_asset::loader::AssetLoader>::load::{{closure}}::hb79ffd15b10bcbfb
                               at /home/hellzbellz/Source/AspenHalls/deps/bevy_aseprite_ultra/src/loader.rs:106:29
```
2 aseprites, 1 that wont crash and 1 that does,
if you ctrl drag the crashing files pixels all the way on canvas it no longer crashes

[crash_files.zip](https://github.com/user-attachments/files/17562919/crash_files.zip)

**Comment by bikeshedder** [comment]
This does indeed seam to be a bug in the `combined_frame_image` function.

Is it okay for you if I include this file in this repository and make a unit test out of it?

**Comment by Hellzbellz123** [comment]
by all means, use it as you will, its a testing asset i made

after some more digging it looks like cells are collections of pixels and can be moved independently of the layer
the combined_frame_image doesnt take into account cells that are off screen? or just negative? 

here is some more data from those files
the cell x/y position is all that appears too change

[struct_debug-crash.txt](https://github.com/user-attachments/files/17563259/struct_debug-crash.txt)
[struct_debug-no-crash.txt](https://github.com/user-attachments/files/17563260/struct_debug-no-crash.txt)

**Comment by bikeshedder** [comment]
I already found and fixed the issue. I'm just cleaning up the code a bit. I like adding a unit test prior to fixing a bug so it never happens again. :+1:

**Comment by bikeshedder** [comment]
I just release `aseprite-loader 0.3.2` on crates.io:

:package: https://crates.io/crates/aseprite-loader/0.3.2

A simple `cargo update` should fix your issue. I replaced the unsafe type conversions by something that can't panic anymore. If the target coordinate is out of bounds the pixel is just ignored.

I guess when @Lommix wrote this code in the first place he didn't think that cels could be clipped. I didn't notice this either when reviewing this code 9 months ago.

**Comment by Hellzbellz123** [comment]
unfortunately bevy_aseprite_ultra hard specifies 0.3.0, so it will have too be patched locally.
however, the first crash test case didnt cover all the issues that existed.

after i updated i reverted my assets too before i removed a bunch of frames and i found a few more cases that caused the same? or maybe a different issue

the test cases are pretty egregious but i think it makes it clear what we dont want

following your earlier example a simple patch
```
    let Some(target_pixel) = target.get_mut(target_index * 4..target_index * 4 + 4)
    else {
        continue;
    };
```
but im not sure if the intended behavior ends up being correct

```
running 1 test
test loader::test_issue_4 ... FAILED

successes:

failures:

---- loader::test_issue_4 stdout ----
thread 'loader::test_issue_4' panicked at src/loader/mod.rs:258:36:
range end index 16388 out of range for slice of length 16384
```


i added a couple more frames too the issue_4 file that dont pass 
[issue_4.zip](https://github.com/user-attachments/files/17564942/issue_4.zip)

**Comment by bikeshedder** [comment]
Argh... I messed that and introduced an off-by-one error. :facepalm: 

It's fixed in the 0.3.3 release:

:package: https://crates.io/crates/aseprite-loader/0.3.3

btw. `bevy-aseprite-ultra` doesn't pin the version to exactly `0.3.0`. The requirement `aseprite-loader = 0.3.0` just means that it has to be at least `0.3.0` but any `0.3.x` version is fine. A `cargo update` should do the trick. No need to patch it locally.
